soc/arm/nordic_nrf/nrf53/soc.c

```diff
@@ -123,6 +134,76 @@ static bool nrf53_anomaly_160_check(void)
 	return true;
 }
 
+#if defined(CONFIG_SOC_NRF53_RTC_PRETICK) && defined(CONFIG_SOC_NRF5340_CPUNET)
+
+BUILD_ASSERT(!IS_ENABLED(CONFIG_WDT_NRFX),
+	     "For CONFIG_SOC_NRF53_RTC_PRETICK watchdog is used internally for the pre-tick workaround on nRF5340 cpunet. Application cannot use the watchdog.");
+
+static inline uint32_t rtc_counter_sub(uint32_t a, uint32_t b)
+{
+	return (a - b) & NRF_RTC_COUNTER_MAX;
+}
+
+static bool rtc_ticks_to_next_event_get(NRF_RTC_Type * rtc, uint32_t selected_cc_mask, uint32_t cntr, uint32_t * ticks_to_next_event)
+{
+	bool result = false;
+
+	/* Let's preload registers to speed-up. The function is not thread-safe anyway. */
+	uint32_t reg_intenset = rtc->INTENSET;
+
+	/* If corresponding EVTEN bit is zero, then interrupt won't be generated as the event itself wont be generated. */
+	reg_intenset &= rtc->EVTEN;
+
+	/* Note: TICK event not handled. */
+
+	if (reg_intenset & NRF_RTC_INT_OVERFLOW_MASK) {
+		/* Overflow can generate an interrupt. */
+		*ticks_to_next_event = NRF_RTC_COUNTER_MAX + 1U - cntr;
+		result = true;
+	}
+
+	for (uint32_t chan = 0; chan < NRF_RTC_CC_COUNT_MAX; chan++) {
+		if ((selected_cc_mask & (1U << chan)) &&
+		    (reg_intenset & NRF_RTC_CHANNEL_INT_MASK(chan))) {
+			/* The CC is in selected mask and is can generate an interrupt. */
+			uint32_t cc = nrf_rtc_cc_get(rtc, chan);
+			uint32_t ticks_to_fire = rtc_counter_sub(cc, cntr);
+
+			if (ticks_to_fire == 0U) {
+				/* When ticks_to_fire == 0, the event should have been just generated
+				 * the interrupt can be already handled or be pending.
+				 * However the next event is expected to be after counter wraps.
+				 */
+				ticks_to_fire = NRF_RTC_COUNTER_MAX + 1U;
+			}
+
+			if (!result) {
+				*ticks_to_next_event = ticks_to_fire;
+				result = true;
+			} else if (ticks_to_fire < *ticks_to_next_event) {
+				*ticks_to_next_event = ticks_to_fire;
+				result = true;
+			}
+			else {
+				/* CC that fires no earlier than already found. */
+			}
+		}
+	}
+
+	return result;
+}
+
+static void rtc_counter_synchronized_get(NRF_RTC_Type * rtc_a, NRF_RTC_Type * rtc_b, uint32_t * counter_a, uint32_t * counter_b)
```

Comment:
Didn't we want this function to be availble to other modules?

Reply:
That was my idea first, however it is required for pretick only, and I have no idea where to put it to allow it to be reused.